File: C#/uri-1033.cs
// URI - 1033: https://www.urionlinejudge.com.br/judge/pt/problems/view/1033

using System;

namespace URI
{
    class Program
    {
        static void Main(string[] args)
        {
            string[] testCase = new string[2];
            long number, result;
            int baseNumber;
            while (true)
            {
                testCase = Console.ReadLine().Split(' ');
                number = long.Parse(testCase[0]);
                baseNumber = int.Parse(testCase[1]);
                if (number == 0 && baseNumber == 0)
                    break;
                result = ((2 * BigMod(number + 1, baseNumber)) - 1 + baseNumber) % baseNumber;
                ShowResult(number, baseNumber, result);
                Counter.caseCounter++;
            }
        }

        private static long BigMod(long number, int baseNumber)
        {
            if (number <= 0)
                return 0;
            long a = 1, b = 0, c = 0, d = 1, r;
            --number;
            while (number > 0)
            {
                if (number % 2 == 1)
                {
                    r = ((d * b) + (c * a)) % baseNumber;
                    b = (d * (b + a) + (c * b)) % baseNumber;
                    a = r;
                }
                r = (c * c + d * d) % baseNumber;
                d = (d * (2 * c + d)) % baseNumber;
                c = r;
                number = number / 2;
            }
            return (a + b) % baseNumber;
        }

        private static void ShowResult(long number, int baseNumber, long result)
        {
            Console.WriteLine($"Case {Counter.caseCounter}: {number} {baseNumber} {result}");
        }
    }

    class Counter
    {
        public static int caseCounter = 1;
    }
}

File: Ruby/uri-1033.rb
# frozen_string_literal: true

# URI - 1033: https://www.urionlinejudge.com.br/judge/pt/problems/view/1033

def calculate_the_number_of_fibonacci_recursive_calls(number_fibonacci, recursive_call_count_array)
  return '1' if number_fibonacci < 2
  return recursive_call_count_array[number_fibonacci] if recursive_call_count_array.size > number_fibonacci

  fibonacci = recursive_call_count_array.size
  while fibonacci <= number_fibonacci
    recursive_call_count_array << 1 + (recursive_call_count_array[fibonacci - 1] +
                                      recursive_call_count_array[fibonacci - 2])
    fibonacci += 1
  end
  recursive_call_count_array[number_fibonacci]
end

recursive_call_count_array = [1, 1]
case_counter = 1
loop do
  input = gets.chomp.split(' ')
  break if input[0] == '0' && input[1] == '0'

  fibonacci_number = input[0].to_i
  base_number = input[1].to_i
  case_info = "Case #{case_counter}: #{fibonacci_number} #{base_number} "
  puts case_info + (calculate_the_number_of_fibonacci_recursive_calls(fibonacci_number,
                                                                      recursive_call_count_array) % base_number).to_s
  case_counter += 1
end
